lib/stats/ManifestPlugin.js

```diff
@@ -0,0 +1,222 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Haijie Xie @hai-x
+*/
+
+"use strict";
+
+const path = require("path");
+const { RawSource } = require("webpack-sources");
+const Compilation = require("../Compilation");
+const HotUpdateChunk = require("../HotUpdateChunk");
+const createSchemaValidation = require("../util/create-schema-validation");
+
+/** @typedef {import("../Compiler")} Compiler */
+/** @typedef {import("../Chunk")} Chunk */
+/** @typedef {import("../Module")} Module */
+/** @typedef {import("../NormalModule")} NormalModule */
+/** @typedef {import("../config/defaults").WebpackOptionsNormalizedWithDefaults} WebpackOptions */
+
+/** @typedef {import("../../declarations/plugins/ManifestPlugin").ManifestPluginOptions} ManifestPluginOptions */
+
+/**
+ * @typedef {object} Asset
+ * @property {string} name
+ * @property {string} path
+ */
+
+/**
+ * @typedef {{ name: string, stage: number }} TapOptions
+ */
+
+/** @type {TapOptions} */
+const TAP_OPTIONS = {
+	name: "ManifestPlugin",
+	stage: Compilation.PROCESS_ASSETS_STAGE_REPORT
+};
+
+const validate = createSchemaValidation(
+	require("../../schemas/plugins/ManifestPlugin.check"),
+	() => require("../../schemas/plugins/ManifestPlugin.json"),
+	{
+		name: "ManifestPlugin",
+		baseDataPath: "options"
+	}
+);
+
+/**
+ * @param {string} filename filename
+ * @returns {string} extname
+ */
+const extname = (filename) => {
+	const replaced = filename.replace(/\?.*/, "");
+	const split = replaced.split(".");
+	const last = split.pop();
+	if (!last) return "";
+	return last && /^(gz|map)$/i.test(last) ? `${split.pop()}.${last}` : last;
+};
+
+class ManifestPlugin {
+	/**
+	 * @param {ManifestPluginOptions} options options
+	 */
+	constructor(options) {
+		validate(options);
+
+		const defaultOptions = {
+			filename: "manifest.json"
+		};
+
+		/** @type {Required<ManifestPluginOptions>} */
+		this.options = Object.assign(defaultOptions, options);
+	}
+
+	/**
+	 * Apply the plugin
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		/** @type {Map<string,Module>} */
+		const moduleAssets = new Map();
```

Comment:
Let's use a WeakMap here - `WeakMap<Compilation, Map<string, Module>>`, so we don't need to clear this because no compilation - no map